runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamPollingScheduler.java

```diff
@@ -0,0 +1,309 @@
+            } catch (Throwable t) {
+                log.error("Unexpected throwable!", t);
+            }
+        }
+    }
+
+    private void tick() {
+        scheduler.submit(new Tick());
+    }
+
+    public void addTask(@Nonnull StreamListener streamListener, @Nonnull String namespace,
+                        @Nonnull String streamTag, @Nonnull List<String> tablesOfInterest,
+                        long lastAddress, int bufferSize) {
+
+        Preconditions.checkArgument(bufferSize >= pollThreshold);
+        synchronized (allTasks) {
+            if (allTasks.containsKey(streamListener)) {
+                // Multiple subscribers subscribing to same namespace and table is allowed
+                // as long as the hashcode() and equals() method of the listeners are different.
+                throw new StreamingException(
+                        "StreamingManager::subscribe: listener already registered " + streamListener);
+            }
+            StreamingTask task = new StreamingTask(runtime, workers, namespace, streamTag, streamListener,
+                    tablesOfInterest, lastAddress, bufferSize);
+            allTasks.put(streamListener, task);
+            log.debug("addTask: added {} {} {}", streamListener, namespace, streamTag);
+            //allTasks.notify();
+            allTasks.notifyAll();
+        }
+    }
+
+    public void removeTask(@Nonnull StreamListener streamListener) {
+        synchronized (allTasks) {
+            allTasks.remove(streamListener);
+            allTasks.notifyAll();
+        }
+    }
+
+    // hole filling mitigation
+    // sequencer regression mitigation
+    // fairness?
+
+    private void waitForTasks() throws InterruptedException {
+        synchronized (allTasks) {
+            while (allTasks.isEmpty()) {
+                allTasks.wait();
+            }
+        }
+    }
+
+    private List<StreamingTask> getTasks(StreamStatus status, Predicate<StreamingTask> predicate) {
+        List<StreamingTask> result = new ArrayList<>();
+        for (StreamingTask task : allTasks.values()) {
+            if (task.getStatus() == status && predicate.test(task)) {
+                result.add(task);
+            }
+        }
+        return result;
+    }
+
+    private List<StreamingTask> getTasks(StreamStatus status) {
+        return getTasks(status, t -> true);
+    }
+
+    private void poll(List<StreamingTask> tasks) {
+        log.info("polling tasks {}", tasks);
+
+        List<List<StreamingTask>> batches = Lists.partition(tasks, pollBatchSize);
+
+        for (List<StreamingTask> batch : batches) {
+            List<StreamAddressRange> pollRequest = new ArrayList<>();
+            for (StreamingTask task : batch) {
+                DeltaStream stream = task.getStream();
+                pollRequest.add(new StreamAddressRange(stream.getStreamId(), Address.MAX, stream.getMaxAddressSeen()));
+            }
+
+            List<StreamAddressRange> coalesced = coalesce(pollRequest);
```

Comment:
Coalesce should be performed before batching otherwise there could be stream queries across batches that aren't coalesced.

Reply:
Done.